src/server/registry/catalog.go

```diff
@@ -68,16 +75,50 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
 		r.sendResponse(w, req, repoNames)
 		return
 	}
-	for _, repo := range repoRecords {
-		valid, err := r.validateRepo(req.Context(), repo.RepositoryID)
-		if err != nil {
-			lib_http.SendError(w, err)
-			return
-		}
-		if valid {
-			repoNames = append(repoNames, repo.Name)
+
+	repoNamesChan := make(chan string)
+	done := make(chan bool, 1)
+	wg.Add(len(repoRecords))
+
+	go func() {
+		defer func() {
+			done <- true
+		}()
+
+		for {
+			select {
+			case str, ok := <-repoNamesChan:
+				if !ok {
+					return
+				}
+				repoNames = append(repoNames, str)
+			}
 		}
+	}()
+
+	for _, repo := range repoRecords {
+		r.wp.GetWorker()
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				r.wp.ReleaseWorker()
+				wg.Done()
+			}()
+
+			valid, err := r.validateRepo(orm.Clone(req.Context()), repo.RepositoryID)
```

Comment:
We can call the tag manager directly to check whether the repository contains tags or not without listing artifacts
Further more, how about to introduce a function in the repository manager to get the repositories with tags directly? We can join the table repository and tag in database level and get the desired result

Reply:
Agree to call the tag manager for checking. Not agree to do it in the repo manager using join between the repo table with the tag table.

Join is an expensive operation on large tables.

Reply:
use the tagCtl for instead.